libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -510,3 +680,95 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
 		);
 
 }
+
+void ContractLevelChecker::checkAmbiguousOverrides(ContractDefinition const& _contract) const
+{
+	vector<FunctionDefinition const*> contractFuncs = _contract.definedFunctions();
+
+	FunctionSet const& baseSet = getBaseFunctions(&_contract);
+	//auto const resolvedBases = resolveBaseContracts(_contract);
+
+	for (auto it = baseSet.cbegin(); it != baseSet.cend(); it = baseSet.upper_bound(*it))
+	{
+		auto [begin,end] = baseSet.equal_range(*it);
+
+		// Only one function
+		if (next(begin) == end)
+			continue;
+
+		// Not an overridable function
+		if (!(*it)->isOverridable())
+			continue;
+
+		// Only look at our direct bases
+		//if (!contains(resolvedBases, (*it)->annotation().contract))
+		//	continue;
+
+		// Function has been explicitly overriden
+		if (contains_if(
+				contractFuncs,
+				[&] (FunctionDefinition const* _f) {
+					return hasEqualNameAndParameters(*_f, **it);
+				}
+			)
+	   	)
+			continue;
+
+		SecondarySourceLocation ssl;
+
+		for (;begin != end; begin++)
+			ssl.append("Definition here: ", (*begin)->location());
+
+		m_errorReporter.typeError(
+			_contract.location(),
+			ssl,
+			"Functions of the same name " +
+			(*it)->name() +
+			" and parameter types defined in two or more base contracts must be overridden in the derived contract."
+		);
+	}
+}
+
+vector<ContractDefinition const*> ContractLevelChecker::resolveOverrideList(OverrideSpecifier const& _overrides) const
+{
+	vector<ContractDefinition const*> resolved;
+
+	for (ASTPointer<UserDefinedTypeName> const& override: _overrides.overrides())
+	{
+		Declaration const* decl  = override->annotation().referencedDeclaration;
+		solAssert(decl, "Expected declaration to be resolved.");
+
+		// If it's not a contract it will be caught
+		// in the reference resolver
+		if (ContractDefinition const* contract = dynamic_cast<decltype(contract)>(decl))
+			resolved.emplace_back(contract);
+	}
+
+	return resolved;
+}
+
+
+ContractLevelChecker::FunctionSet const& ContractLevelChecker::getBaseFunctions(ContractDefinition const* _contract) const
+{
+	auto const& result = m_contractBaseFunctions.find(_contract);
+
+	if (result != m_contractBaseFunctions.cend())
+		return result->second;
+
+	FunctionSet set, tmpSet;
+
+	for (auto const* base: resolveBaseContracts(*_contract))
+	{
+		for (auto const* func: base->definedFunctions())
+			tmpSet.insert(func);
+
+		for (auto const& func: getBaseFunctions(base))
+			if (tmpSet.find(func) == set.cend())
```

Comment:
With C++20 we could use ``tmpSet.contains(func)`` here :-). @chriseth usually prefers ``tmpSet.count(func)`` instead, but that will be slightly more complex here since it's a multiset... but the fact that it should definitely be ``tmpSet.cend()`` instead of ``set.cend()`` indicates that simplifying to ``tmpSet.count(func)`` might still be worth it ;-).